WDL basic support.  Alpha testing phase.

DailyDreaming opened the pull request:
Support for .wdl files currently working for the GATK tutorials 1-4 on the WDL webpage.  Have not tested on any other files, so currently this is still in alpha.

cgcloud-jenkins commented:
Build finished. 8 tests run, 0 skipped, 0 failed.

cgcloud-jenkins commented:
Build finished. 666 tests run, 103 skipped, 5 failed.

benedictpaten commented:
Hey Lon, can you rebase on master so that we can see the diffs cleanly? Also, holy shit, how can this 180K lines? Are you checking in some big binary files? I'd like to avoid that if at all possible.

benedictpaten commented:
Finally, going to look closely at the tests on this too.

joelarmstrong commented:
The 100k+ extra lines are VCF test data for the Broad's example workflows I think. It shouldn't be a big deal IMO because it'll compress down.

DailyDreaming commented:
I haven't pushed any of the new stuff.  Just did a rebase to merge back with master (which seems to have messed up).  I'm cleaning it up and will have the new stuff up soon.

cgcloud-jenkins commented:
Build finished. 8 tests run, 0 skipped, 0 failed.

cgcloud-jenkins commented:
Build finished. 666 tests run, 103 skipped, 5 failed.

diekhans commented:
It might also make sense to use git lfs for the large test file. 

Joel Armstrong <notifications@github.com> writes:
> The 100k+ extra lines are VCF test data for the Broad's example workflows I
> think. It shouldn't be a big deal IMO because it'll compress down.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.*
>

DailyDreaming commented:
Compressed the big test files.  Added a few more tests (would like to add even more, though for now the workflows are a kind of blunt object testing many params at once).  Added commenting, which hopefully makes it more understandable.  Just realized I should update the documentation, and move the tests to the test directory.  Feedback very welcome.  Apologies if I did anything dodgy in the code.  Many sections are stubs that should eventually be fleshed out to include more variables and types.

cgcloud-jenkins commented:
Build finished. 28 tests run, 0 skipped, 20 failed.

cgcloud-jenkins commented:
Build finished. 724 tests run, 102 skipped, 56 failed.

cgcloud-jenkins commented:
Build finished. 8 tests run, 0 skipped, 0 failed.

cgcloud-jenkins commented:
Build finished. 8 tests run, 0 skipped, 0 failed.

cgcloud-jenkins commented:
Build finished. 677 tests run, 102 skipped, 9 failed.

cgcloud-jenkins commented:
Build finished. 677 tests run, 102 skipped, 9 failed.

cgcloud-jenkins commented:
Build finished. 8 tests run, 0 skipped, 0 failed.

cgcloud-jenkins commented:
Build finished. 677 tests run, 102 skipped, 8 failed.

cgcloud-jenkins commented:
Build finished. 8 tests run, 0 skipped, 0 failed.

cgcloud-jenkins commented:
Build finished. 677 tests run, 102 skipped, 4 failed.

cgcloud-jenkins commented:
Build finished. 8 tests run, 0 skipped, 0 failed.

cgcloud-jenkins commented:
Build finished. 677 tests run, 102 skipped, 4 failed.

ejacox commented:
Jenkins, skip this test

DailyDreaming commented:
jenkins skip this test in commit

DailyDreaming commented:
Jenkins, skip this test please.

cgcloud-jenkins commented:
Build finished. 8 tests run, 0 skipped, 0 failed.

cgcloud-jenkins commented:
Build finished. 8 tests run, 0 skipped, 0 failed.

cgcloud-jenkins commented:
Build finished. 8 tests run, 0 skipped, 0 failed.

cgcloud-jenkins commented:
Build finished. 677 tests run, 102 skipped, 4 failed.

cgcloud-jenkins commented:
Build finished. 677 tests run, 102 skipped, 4 failed.

cgcloud-jenkins commented:
Build finished. 677 tests run, 102 skipped, 4 failed.

cgcloud-jenkins commented:
Build finished. 8 tests run, 0 skipped, 0 failed.

cgcloud-jenkins commented:
Build finished. 677 tests run, 102 skipped, 4 failed.

cgcloud-jenkins commented:
Build finished. 8 tests run, 0 skipped, 0 failed.

cgcloud-jenkins commented:
Build finished. 677 tests run, 102 skipped, 4 failed.

DailyDreaming commented:
The GATK tests were failing on Jenkins because Jenkins has Java 6, and GATK requires Java 7.  These have since been moved to integration tests, as they are long and we should initiate a longer, more thorough testing for release, but not necessarily on Jenkins.  The ENCODE workflow test will still run on jenkins and should test most of the wdl compilation features in addition to the other smaller tests.

cgcloud-jenkins commented:
Build finished. 8 tests run, 0 skipped, 0 failed.

cgcloud-jenkins commented:
Build finished. 8 tests run, 0 skipped, 0 failed.

cgcloud-jenkins commented:
Build finished. 673 tests run, 102 skipped, 0 failed.

cgcloud-jenkins commented:
Build finished. 673 tests run, 102 skipped, 0 failed.

DailyDreaming commented:
All tests are passing, but still have tiny x's.  >.>

DailyDreaming commented:
Jenkins, test this please.

cgcloud-jenkins commented:
Build finished. 8 tests run, 0 skipped, 1 failed.

cgcloud-jenkins commented:
Build finished. 673 tests run, 102 skipped, 0 failed.

DailyDreaming commented:
Jenkins, test this please.

cgcloud-jenkins commented:
Build finished. 8 tests run, 0 skipped, 1 failed.

DailyDreaming commented:
Jenkins, test this please.

cgcloud-jenkins commented:
Build finished. 8 tests run, 0 skipped, 1 failed.

cgcloud-jenkins commented:
Build finished. 673 tests run, 102 skipped, 0 failed.

DailyDreaming commented:
Jenkins, test this please.

cgcloud-jenkins commented:
Build finished. 8 tests run, 0 skipped, 0 failed.

cgcloud-jenkins commented:
Build finished. 673 tests run, 102 skipped, 0 failed.

ejacox commented on src/toil/wdl/toilwdl.py:
```diff
@@ -0,0 +1,2077 @@
+        :return: A string containing the apiDockerCall() that will run the job.
+        '''
+        fn_section = ''
+        fn_section = fn_section + '    apiDockerCall(job, image
```
Where is this called?

DailyDreaming replied:
Made it an option.

ejacox commented on src/toil/wdl/toilwdl.py:
```diff
@@ -0,0 +1,2077 @@
+        :return: A string containing the apiDockerCall() that will run the job.
+        '''
+        fn_section = ''
+        fn_section = fn_section + '    apiDockerCall(job, image
```
Is this called?

DailyDreaming replied:
Made it an option.

ejacox commented on src/toil/wdl/toilwdl.py:
```diff
@@ -0,0 +1,2077 @@
+        :return: A string containing the apiDockerCall() that will run the job.
+        '''
+        fn_section = ''
+        fn_section = fn_section + '    apiDockerCall(job, image
```
should be 'with open'

DailyDreaming replied:
Fixed.

ejacox commented on src/toil/wdl/toilwdl.py:
```diff
@@ -0,0 +1,2077 @@
+        :return: Creates the self.tasks_dictionary necessary for much of the
+        parser.  Returning it is only necessary for unittests.
+        '''
+        wdl = open(self.wdl_file, 'r')
```
Should be 'with open'

DailyDreaming replied:
Fixed.

ejacox commented on src/toil/wdl/toilwdl.py:
```diff
@@ -0,0 +1,2077 @@
+        :return: Creates the self.workflows_dictionary necessary for much of the
+        parser.  Returning it is only necessary for unittests.
+        '''
+        wdl = open(self.wdl_file, 'r')
```
Should be 'with open'

DailyDreaming replied:
Fixed.

ejacox commented on src/toil/wdl/toilwdl.py:
```diff
@@ -0,0 +1,2077 @@
+        :return: A string containing the apiDockerCall() that will run the job.
+        '''
+        fn_section = ''
+        fn_section = fn_section + '    apiDockerCall(job, image
```
This needs to parse the Toil options.

DailyDreaming replied:
Fixed.

ejacox commented on src/toil/wdl/toilwdl.py:
```diff
@@ -0,0 +1,2077 @@
+        :return: A string containing the apiDockerCall() that will run the job.
+        '''
+        fn_section = ''
+        fn_section = fn_section + '    apiDockerCall(job, image
```
Both of these create_*_dict functions read and parse the WDL file. This can be done once and the parser reused.

DailyDreaming replied:
I could and originally did.  I liked the isolation though, though I suppose it's bad form.  I'll look at changing it.

DailyDreaming replied:
Fixed.

ejacox commented on src/toil/wdl/toilwdl.py:
```diff
@@ -0,0 +1,2077 @@
+import sys
+import json
+import logging
+
```
A paragraph describing the general design would be useful here.

DailyDreaming replied:
Moved the class up and the two orphaned defs down so that it's more visible.  Hope it's enough detail for a general overview.

ejacox commented on src/toil/wdl/toilwdl.py:
```diff
@@ -0,0 +1,2077 @@
+        :return: A string containing the apiDockerCall() that will run the job.
+        '''
+        fn_section = ''
+        fn_section = fn_section + '    apiDockerCall(job, image
```
Shouldn't this delete the output_file afterwards? An option to keep it would be useful for debugging.

DailyDreaming replied:
Fixed.

ejacox commented on src/toil/wdl/toilwdl.py:
```diff
@@ -0,0 +1,2077 @@
+
+        # a job's 'level' on the DAG
+        self.task_priority = 0
+
```
module_string and main_prelude string should be class attributes

DailyDreaming replied:
I've made module_string into its own def in case this wants to be handled dynamically in the future and incorporated the other as a def supporting the main write_main() def.

cgcloud-jenkins commented:
Build finished. 8 tests run, 0 skipped, 0 failed.

cgcloud-jenkins commented:
Build finished. 8 tests run, 0 skipped, 0 failed.

cgcloud-jenkins commented:
Build finished. 8 tests run, 0 skipped, 0 failed.

cgcloud-jenkins commented:
Build finished. 722 tests run, 23 skipped, 1 failed.

cgcloud-jenkins commented:
Build finished. 722 tests run, 23 skipped, 8 failed.

cgcloud-jenkins commented:
Build finished. 722 tests run, 23 skipped, 13 failed.

cgcloud-jenkins commented:
Build finished. 8 tests run, 0 skipped, 0 failed.

cgcloud-jenkins commented:
Build finished. 722 tests run, 23 skipped, 0 failed.

jessebrennan commented on docs/running/wdl.rst:
```diff
@@ -0,0 +1,139 @@
+WDL Support in Toil
+********
```
Build errors for `make docs` complain about `WARNING: Title underline too short.`

This happens lots of times for all of the titles.

DailyDreaming replied:
Fixed.

jessebrennan commented on docs/running/wdl.rst:
```diff
@@ -0,0 +1,139 @@
+WDL Support in Toil
```
Other docs building error:
`checking consistency... /Users/jesse/Documents/genomics/toil/docs/running/wdl.rst:: WARNING: document isn't included in any toctree`

This means the only way to access the WDL docs is to manually navigate to toil/docs/_build/html/running/wdl.html

DailyDreaming replied:
Added to index.rst.

jessebrennan commented on docs/running/wdl.rst:
```diff
@@ -0,0 +1,139 @@
+https://github.com/ENCODE-DCC/pipeline-container
+
+We've included the wdl file and data files in the toil repository needed to run this example.  First, find the wdl file:
+'src/toil/test/wdl/wdl_templates/testENCODE/encode_mapping_workflow.wdl'
```
It would be easier for the user to just link the downloadable file as is done for the sort example

``Download :download:`the example code <../../src/toil/test/sort/sort.py>`.``

DailyDreaming replied:
Amended, though it seems it'll need to be merged before I can see it properly on the docs?

jessebrennan commented on docs/running/wdl.rst:
```diff
@@ -0,0 +1,139 @@
+How to Run a WDL file in Toil
+-----------
+Recommended best practice when running wdl files is to first use the Broad's wdltool for syntax validation and generating the
+needed json input file.  Full documentation can be found on the repository_, and a precompiled jar binary can be downloaded here: wdltool_.
```
Clearly Java is a requirement. This  should be more explicit with the version needed to run this successfully.

DailyDreaming replied:
Fixed.

jessebrennan commented on docs/running/wdl.rst:
```diff
@@ -0,0 +1,139 @@
+https://github.com/ENCODE-DCC/pipeline-container
+
+We've included the wdl file and data files in the toil repository needed to run this example.  First, find the wdl file:
+'src/toil/test/wdl/wdl_templates/testENCODE/encode_mapping_workflow.wdl'
```
Also this file doesn't exist for me. Instead there is just a `wdl_templates.zip`. Should I just unzip it or does this happen automatically after I've run tests or what?

DailyDreaming replied:
I've amended this.  Let me know what you think.

jessebrennan commented on docs/running/wdl.rst:
```diff
@@ -0,0 +1,139 @@
+found here:
+
+src/toil/test/wdl/ENCODE_data/reference/GRCh38_chr21_bwa.tar.gz
+src/toil/test/wdl/ENCODE_data/ENCFF000VOL_chr21.fq.gz
```
A user won't have access to these files unless they have installed toil from source and this isn't explained in the  docs here. Other options are to just link to downloadable copies or just tell them to build from source.

DailyDreaming replied:
Again, I've amended this.  Let me know what you think!

cgcloud-jenkins commented:
Build finished. 8 tests run, 0 skipped, 0 failed.

cgcloud-jenkins commented:
Build finished. 722 tests run, 23 skipped, 0 failed.

joelarmstrong commented:
Well, I broke github somehow--some of my disappearing comments reappeared as "outdated", but now a lot of my comments all ended up in the same (wrong) place, just after `command_var_decl = 'command' + str(self.command_number) `. I'll try to move them to the proper place tomorrow.

joelarmstrong commented on src/toil/wdl/toilwdl.py:
```diff
@@ -0,0 +1,2098 @@
+            cmd_name = cmd[0]
+            cmd_type = cmd[1]
+            cmd_actions_dict = cmd[2]
+     
```
You'll deadlock here if the process outputs more than ~4KB of data to stdout. The [Python docs](https://docs.python.org/2/library/subprocess.html#subprocess.Popen.wait) have a little warning about this, but it should be much more prominent. The core issue is that pipes have a limited buffer size, and if the buffer is full, any writes will block until the other side reads from the pipe and frees up buffer space.

You should use `this_process.communicate()` and just throw away the data.

joelarmstrong replied:
nit: again, os.path.basename is probably a better idea here.

joelarmstrong replied:
May want to document here (or somewhere else) that this won't work with `${default="foo" bar}` or `${true="foo" false="bar" Boolean baz}`.

joelarmstrong commented:
Lol, OK, clearly github has a bug that won't let me comment past line 1417, so I can't move those comments to the proper place. Maybe I should prepend the line number to the comments as a workaround?

DailyDreaming commented:
Thanks again for all of the comments Joel.  Really really appreciate the time you put into this.  If there are many more, I'd be happy to get them via email/slack or (haha) prepended with lines here.  Thanks again, I'll try to correct these by Monday!

joelarmstrong commented:
Here are the broken comments, this time prefixed by line numbers:

- [Line 1468](https://github.com/BD2KGenomics/toil/pull/1823/files/7ec2380eb5e09ad724179ea7c7faa651420093de#diff-79452574b4265a04dfebb2f477281c16R1468):

You'll deadlock here if the process outputs more than ~4KB of data to stdout. The [Python docs](https://docs.python.org/2/library/subprocess.html#subprocess.Popen.wait) have a little warning about this, but it should be much more prominent. The core issue is that pipes have a limited buffer size, and if the buffer is full, any writes will block until the other side reads from the pipe and frees up buffer space.

You should use `this_process.communicate()` and just throw away the data.

- [Line 1503](https://github.com/BD2KGenomics/toil/pull/1823/files/7ec2380eb5e09ad724179ea7c7faa651420093de#diff-79452574b4265a04dfebb2f477281c16R1503):

nit: again, os.path.basename is probably a better idea here.

- [Line 1594](https://github.com/BD2KGenomics/toil/pull/1823/files/7ec2380eb5e09ad724179ea7c7faa651420093de#diff-79452574b4265a04dfebb2f477281c16R1594):

May want to document here (or somewhere else) that this won't work with `${default="foo" bar}` or `${true="foo" false="bar" Boolean baz}`.

- [Line 1625](https://github.com/BD2KGenomics/toil/pull/1823/files/7ec2380eb5e09ad724179ea7c7faa651420093de#diff-79452574b4265a04dfebb2f477281c16R1625) (basically just commenting on the `translate_wdl_string_to_python_string` function as a whole):

I'm not so sure about this whole function. What if the string contains a `'` character? It seems like the above could be replaced with something like (roughly) `output_string = "'" + some_string.replace('${', "' + ").replace('}', " + '") + "'"`.

Escape characters like `\n` will also get interpreted by python--that's not really a blocker for an experimental alpha, but something to keep in mind.

DailyDreaming commented on integration_tests/testIntegration.py:
```diff
@@ -0,0 +1,160 @@
+
+#######################################
+# Note: GATK.jar requires java 7      #
+# jenkins has only java 6 (11-22-2017)#
```
Thanks!

DailyDreaming commented on integration_tests/testIntegration.py:
```diff
@@ -0,0 +1,160 @@
+# jenkins has only java 6 (11-22-2017)#
+#######################################
+
+class TestCase(unittest.TestCase):
```
Fixed.

DailyDreaming commented on integration_tests/testIntegration.py:
```diff
@@ -0,0 +1,160 @@
+
+        self.program = os.path.abspath("src/toil/wdl/toilwdl.py")
+        self.output_dir = os.path.abspath("src/toil/wdl/wdl_working/")
+        self.output_file = os.path.abspath("toilwdl_compiled.py")
```
Fixed.

DailyDreaming commented on integration_tests/testIntegration.py:
```diff
@@ -0,0 +1,160 @@
+
+    def tearDown(self):
+        """Default tearDown for unittest."""
+        unittest.TestCase.tearDown(self)
```
Fixed (tempDir now created and deleted).

DailyDreaming commented on integration_tests/testIntegration.py:
```diff
@@ -0,0 +1,160 @@
+
+        :param ref_dir:
+        :return:
+        """
```
Fixed.

DailyDreaming commented on integration_tests/testIntegration.py:
```diff
@@ -0,0 +1,160 @@
+                    for test_file in test_output_files:
+                        if file == test_file:
+                            test_filepath = os.path.join(self.output_dir, file)
+                            print(file)
```
Fixed.

DailyDreaming commented on integration_tests/testIntegration.py:
```diff
@@ -0,0 +1,160 @@
+        :param filepath1:
+        :param filepath2:
+        :return:
+        """
```
Fixed.

DailyDreaming commented on src/toil/test/wdl/toilwdlTest.py:
```diff
@@ -0,0 +1,163 @@
+from toil.wdl.toilwdl import ToilWDL
+import zipfile
+
+class TestCase(unittest.TestCase):
```
Fixed.

DailyDreaming commented on src/toil/test/wdl/toilwdlTest.py:
```diff
@@ -0,0 +1,163 @@
+
+        self.program = os.path.abspath("src/toil/wdl/toilwdl.py")
+        self.output_dir = os.path.abspath("src/toil/wdl/wdl_working/")
+        self.output_file = os.path.abspath("toilwdl_compiled.py")
```
Fixed.

DailyDreaming commented on src/toil/test/wdl/toilwdlTest.py:
```diff
@@ -0,0 +1,163 @@
+        json_dict = t.dict_from_JSON("src/toil/test/wdl/wdl_templates/t01/helloHaplotypeCaller_inputs.json")
+        assert json_dict == default_json_dict_output
+
+    def compare_runs(self, ref_dir):
```
Fixed.

DailyDreaming commented on src/toil/wdl/toilwdl.py:
```diff
@@ -0,0 +1,2077 @@
+import collections
+import subprocess
+import toil.wdl.wdl_parser as wdl_parser
+import sys
```
Fixed.

DailyDreaming commented on src/toil/wdl/toilwdl.py:
```diff
@@ -0,0 +1,2077 @@
+import subprocess
+import toil.wdl.wdl_parser as wdl_parser
+import sys
+import json
```
Fixed.

DailyDreaming commented on src/toil/wdl/toilwdl.py:
```diff
@@ -0,0 +1,2077 @@
+import toil.wdl.wdl_parser as wdl_parser
+import sys
+import json
+import logging
```
Fixed.

DailyDreaming commented on src/toil/wdl/toilwdl.py:
```diff
@@ -0,0 +1,2077 @@
+
+    bashfile_string = bashfile_string + bashfile_suffix
+
+    with open(temp_dir + '/' + job_name + '_script.sh', 'w') as bashfile:
```
Fixed.

DailyDreaming commented on src/toil/wdl/toilwdl.py:
```diff
@@ -0,0 +1,2077 @@
+    github.com/ENCODE-DCC/pipeline-container/blob/master/local-workflows/encode_mapping_workflow.wdl
+
+    Additional support to be broadened to include more features soon.
+    '''
```
Moved to the top with additional explanation.

DailyDreaming commented on src/toil/wdl/toilwdl.py:
```diff
@@ -0,0 +1,2077 @@
+        # unique number for a job
+        self.task_number = 0
+
+        # a job's 'level' on the DAG
```
Thanks!

DailyDreaming commented on src/toil/wdl/toilwdl.py:
```diff
@@ -0,0 +1,2077 @@
+                tsv_array.append(array_variables)
+        return(tsv_array)
+
+    def create_csv_array(self, csv_filepath):
```
Fixed.

DailyDreaming commented on src/toil/wdl/toilwdl.py:
```diff
@@ -0,0 +1,2077 @@
+    def dict_from_JSON(self, JSON_file):
+        '''
+        Takes a WDL-mapped json file and creates a dict containing the bindings.
+        The 'return' value is only used ofr unittests.
```
Haha, this really bothered me too.  Fixed.

DailyDreaming commented on src/toil/wdl/toilwdl.py:
```diff
@@ -0,0 +1,2077 @@
+            var_name, var_type, var_value = self.parse_task_declaration(declaration_subAST)
+            var_truple = (var_name, var_type, var_value)
+            declaration_array.append(var_truple)
+            self.tasks_dictionary.setdefault(task_name, {})['inputs'] = declaration_array
```
Thanks, it's good to be aware of both.  I've never used defaultdict({}) before.

DailyDreaming commented on src/toil/wdl/toilwdl.py:
```diff
@@ -0,0 +1,2077 @@
+                var_type = declaration_subAST.attr("type").source_string
+
+            elif isinstance(declaration_subAST.attr("type"), wdl_parser.Ast):
+
```
Added a bit more explanation that hopefully makes it a little easier to understand.  I'm not aware of all of the cases though.  Once this is more robust, a good overview of the patterns would be good and I'll try to add it.

DailyDreaming commented on src/toil/wdl/toilwdl.py:
```diff
@@ -0,0 +1,2077 @@
+                    raise NotImplementedError
+                if isinstance(var_value_rhs, wdl_parser.AstList):
+                    raise NotImplementedError
+
```
Same as above and added a bit more explanation.

diekhans replied:
split into  more function would make it easier to understand

DailyDreaming commented on src/toil/wdl/toilwdl.py:
```diff
@@ -0,0 +1,2077 @@
+        2. Calls (similar to a python def)
+        3. Scatter (which expects to map to a Call or multiple Calls)
+
+        Returns nothing but creates the self.jobs_dictionary necessary for much
```
Fixed.

DailyDreaming commented on src/toil/wdl/toilwdl.py:
```diff
@@ -0,0 +1,2077 @@
+        workflow_name = workflow.attr('name').source_string
+
+        wf_declared_dict = {}
+        for i in workflow.attr("body"):
```
Fixed.

DailyDreaming commented on src/toil/wdl/toilwdl.py:
```diff
@@ -0,0 +1,2077 @@
+                    else:
+                        raise RuntimeError('Scatter failed.  Scatter collection is not an array.')
+                else:
+                    raise RuntimeError('Scatter failed.  Scatter collection not found in jobs_dictionary.')
```
Fixed.

DailyDreaming commented on src/toil/wdl/toilwdl.py:
```diff
@@ -0,0 +1,2077 @@
+            if aaf in self.workflows_dictionary['scatter_calls']:
+                iterator = self.workflows_dictionary['scatter_calls'][aaf]
+                main_section = main_section + '        ' + aaf + ' = []\n'
+                main_section = main_section + '        ' + aaf + '0 = ' + str(self.tsv_dict[aaf]) + '\n'
```
This is actually searching the tsv_dict for the variable input for the scatter.  It's a bit clunky and obfuscated, and I'll write something to handle this better in beta.

DailyDreaming commented on src/toil/wdl/toilwdl.py:
```diff
@@ -0,0 +1,2077 @@
+                main_section = main_section + '            for i in ' + iterator + '0:\n'
+                main_section = main_section + '                if os.path.isfile(str(i)):\n'
+                main_section = main_section + '                    ' + iterator + '0 = toil.importFile("file://" + os.path.abspath(i))\n'
+                main_section = main_section + '                    ' + iterator + '0_preserveThisFilename = i.split("/")[-1]\n'
```
Fixed.

DailyDreaming commented on src/toil/wdl/toilwdl.py:
```diff
@@ -0,0 +1,2077 @@
+            if skip_first == 0:
+                main_section = main_section + '\n        job0 = job0.encapsulate()\n'
+            skip_first = 0
+        main_section = main_section[:-34]
```
Fixed, but in a different way that I hope is equivalent.

DailyDreaming commented on src/toil/wdl/toilwdl.py:
```diff
@@ -0,0 +1,2098 @@
+        for cmd in self.tasks_dictionary[job_task_reference]['raw_commandline']:
+            cmd_name = cmd[0]
+            cmd_type = cmd[1]
+            cmd_actions_dict = cmd[2]
```
Fixed.  Really like that shortcut.

DailyDreaming commented on src/toil/wdl/wdl_parser.py:
```diff
@@ -0,0 +1,3902 @@
+
```
Fixed.

DailyDreaming commented on src/toil/wdl/toilwdl.py:
```diff
@@ -0,0 +1,2077 @@
+            data = json.load(data_file)
+        for d in data:
+            d_list = d.split('.')
+            self.json_dict[d_list[-1]] = data[d]
```
This is a bug.  It's something that needs to be expanded to include other wdl files as well and means a rework of how the dictionaries are built so I've put it off for now.  But yes, this definitely needs to be addressed in the future.

diekhans commented on src/toil/wdl/toilwdl.py:
```diff
@@ -0,0 +1,2077 @@
+        cmd + '\n' + \
+        ')' + '\n\n' + \
+        '# gather the input command return code' + '\n' + \
+        'echo $? > "$tmpDir/rc.tmp"' + '\n\n'
```
a `heredoc'  as a format string with named parameters would be more robust as well.  In general, one has
to be very careful generating shell code.  It is very easy to create invalid syntax on input. Using shlex.quote() to quote const values before formatting into the text is a good idea.

Also, the bash script doesn't set any error handling options.  Errors will be lost unless a human being reads stderr.  I  include
`set -beEu -o pipefail`
at the top of bash scripts unless there is a really good reason not to.

Robust code is bash is very hard.

DailyDreaming commented:
Thanks for all of the awesome comments Joel, I've learned a lot from your comments and will incorporate `os.path.join`, `os.path.basename`, heredocs, `this_process.communicate()` and some of the other sensible conventions I was unaware into all of my coding, where due, for a long time.  Really appreciate that.

That said, I fixed most of the comments.  I am setting up an ftp on Azure for the datasets, and have not yet finished the heredocs.  It needs a more intelligent DAG and more intelligent parsing of the JSON variables (to include functions with exactly named variables and also multiple wdl/json imports), though I will leave these two for the beta.

Thanks again Joel!

diekhans commented:
We should not zip files.  These are not maintainable.  You can't diff the contents if you need to change them, they are painful to check out for developers.   The test cases include a JAR file and genomic sequences, bam incies, etc.

We need to seriously rethink how to handle these

DailyDreaming commented:
> In src/toil/wdl/toilwdl.py:
> 
> > +        '# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,' + '\n' + \
> +        '# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE' + '\n' + \
> +        '# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE' + '\n\n' + \
> +        '# make a temp directory w/identifier' + '\n' + \
> +        'tmpDir=$(mktemp -d /' + docker_dir + '/execution/tmp.XXXXXX)' + \
> +        '\n' + \
> +        'chmod 777 $tmpDir' + '\n' + \
> +        '# set destination for java to deposit all of its files' + '\n' + \
> +        'export _JAVA_OPTIONS=-Djava.io.tmpdir=$tmpDir' + '\n' + \
> +        'export TMPDIR=$tmpDir' + '\n\n' + \
> +        '(' + '\n' + \
> +        'cd /' + docker_dir + '/execution' + '\n' + \
> +        cmd + '\n' + \
> +        ')' + '\n\n' + \
> +        '# gather the input command return code' + '\n' + \
> +        'echo $? > "$tmpDir/rc.tmp"' + '\n\n'
> 
> a `heredoc' as a format string with named parameters would be more robust as well. In general, one has
> to be very careful generating shell code. It is very easy to create invalid syntax on input. Using shlex.quote() to quote const values before formatting into the text is a good idea.
> 
> Also, the bash script doesn't set any error handling options. Errors will be lost unless a human being reads stderr. I include
> set -beEu -o pipefail
> at the top of bash scripts unless there is a really good reason not to.
> 
> Robust code is bash is very hard.

Thanks Mark!  I appreciate that.  I'll add the code and the `set -beEu -o pipefail` into a heredoc.  I like the additional error handling and the subsequent pipe to stdout.  I'll experiment with this, as I'm also writing a debugging tutorial for the debugging PR.  Thanks again!

joelarmstrong commented:
Reopening!

joelarmstrong commented:
Thanks for the very thorough revision! I just did another quick pass. LGTM, pending removal of the test data, so I'll just switch my review to "approve" preemptively.

Remember when you remove the test data that you have to erase it from previous commits as well (annoyingly). Once you remove the test data, you might want to use a script like https://stackoverflow.com/a/42544963 just to confirm the large blobs are gone from all commits, not just the current commit.

diekhans commented on src/toil/wdl/toilwdl.py:
```diff
@@ -0,0 +1,2125 @@
+                                job = self.parse_workflow_call(j, scatter_num=str(scatter_num))
+                                self.workflows_dictionary.setdefault((self.task_priority, self.task_number, task_being_called, task_alias), {})['job_declarations'] = job
+                            scatter_num = scatter_num + 1
+                    else:
```
this code block is too big for those of use with self-inflicted brain damage to understand.  Again suggest, small well named functions make code easier to understand

DailyDreaming replied:
Thanks Mark!  I broke up those two functions in an effort to save myself from code-induced brain damage in the future.  It's a bad habit that I haven't completely gotten rid of, and thanks for pointing it out!

DailyDreaming commented:
Jenkins, test this please.

DailyDreaming commented:
Branch should be up-to-date with master and all large files removed from all commits.  I modified a few of the strings to .format and I think it looks cleaner.  I can reformat more of them, but otherwise I think this is ready to merge.

benedictpaten commented:
@ejacox @jessebrennan Can you approve / provide more input. Seems like this is done and ready to merge. Would be good to merge in the next couple of days.
